Add Mono EventPipe Sample Profiler support.

lateralusX opened the pull request:
Implement support for EventPipe Sample Profiler on Mono inline with CoreClr Sample Profiler behaviour.  By default CoreClr sample profiler tries to run at 1000 hertz (every ms) and on each sample it will stop runtime, snap callstacks for all managed threads, submit EventPipe sample profile events and resume runtime. Doing a full stop/restart of the runtime on every sample could be quite invasive, but is probably the most portable way of implementing it working on most platforms.

This PR implements the same logic, but on Mono runtime, stopping the runtime, record all callstacks for all managed threads that should be sampled, restart runtime and then write all sample profile events into EventPipe. Note that events are written after the runtime has resumed since all code executed when runtime is stopped needs to be async safe (needed when running in preemptive mode) so we can't call into EventPipe at that point.

Going forward we should investigate alternative ways to do sample profiling depending on underlying platform and OS support. Currently implementation will work on all supported platforms, but it will not be as accurate as it could be (especially when using safe points and coop enabled runtime) and impacts measured target. Mono's profiler uses Signals/SuspendThread and for platforms supporting these API's, that could be an alternative implementation. It could also be worth to look into CPU hardware counters using ETW kernel log session on Windows and perf_event_open on Linux.

lateralusX committed:
Generic suspend/resume of runtime.

lateralusX committed:
Implement EventPipe sample profiler on Mono.

lateralusX committed:
Reworked callback and naming.

lateralusX committed:
Exclude managed wrapper methods in callstacks.

lateralusX committed:
Fix build error.

lambdageek reviewed (approved):
LGTM!

naricc commented on src/mono/mono/metadata/gc-internals.h:
```diff
@@ -430,4 +430,11 @@ extern gboolean mono_do_not_finalize;
+ * Unified runtime stop/restart world, SGEN Only.
+ * Will take and release the LOCK_GC.
+ */
+void mono_stop_world (MonoThreadInfoFlags flags);
```
Why wouldn't mono_gc_stop_world work here? I see that this calls unified_suspend_stop_world, but I am not sure I appreciate what the differences are in the different ways to stop the world.

lateralusX replied:
mono_gc_stop_world currently have a number of side effects, mostly stats/events and logging, but also doing some pre-work on togglerefs etc, so its not a pure STW feature and using that from the sample profiler, that does this at very high frequency will cause to much side effects so we needed a cleaner STW, closer to unified_suspend_stop_world. We also need the ability to better control the exclude threads compared to what's done during regular GC STW. I currently made mono_stop_world/mono_restart_world internal to the runtime and in the long run I would like the code to move out from SGEN into a runtime specific feature (STW is not GC specific feature). I guess we could potentially re-wire mono_gc_stop_world into just calling internal mono_stop_world, but if we like mono_gc_stop_world to be close to regular GC STW then I believe we should keep it as is.

lambdageek replied:
@lateralusX `mono_gc_stop_world` is very new and has only one caller that has nothing to do with GC.  It was added basically just to add an easy to use STW api.  If you update its only caller to use `mono_stop_world`, you can get rid of it.

lateralusX replied:
OK, cool, so we could then remove mono_gc_stop_world/mono_gc_restart_world from mono-gc.h and switch external GC call sites into using mono_stop_world/mono_start_world. Maybe we should put those functions in some other internal runtime header? @lambdageek any suggestions?

lambdageek replied:
@lateralusX `threads-types.h`, maybe. Although I don't feel particularly strongly about it one way or another.

lateralusX replied:
Replaced mono_gc_stop|restart_world with mono_stop|restart_world. Kept in gc-internals.h for now until we move generic STW code out of GC codebase.

lateralusX committed:
Review feedback

Replace mono_gc_stop|restart_world with mono_stop|restart_world.
Keep in gc-internals.h until we move STW code into none GC dependent code.
